KAFKA-6119: Bump epoch when expiring transactions in the TransactionCoordinator

apurvam opened the pull request:
A description of the problem is in the JIRA. I have added an integration test which reproduces the original scenario, and also added unit test cases.

Commit:
Bump the epoch when we abort a transaction on the coordinator

Commit:
Lock the transaction metadata before fencing the epoch. Make the case matching exhaustive

apurvam commented:
cc @hachikuji @guozhangwang

apurvam commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -434,21 +435,48 @@ class TransactionCoordinator(brokerId: Int,
+            s"to error: $error")
+        case Right(epochAndMetadata) =>
+          val txnMetadata = epochAndMetadata.transactionMetadata
+          txnMetadata.inLock(txnMetadata.fenceProducerEpoch())
```
I have emulated the locking strategy in `TransactionCoordinator.handleInitProducerId` where the fencing and the update to the log happen with different lock acquisitions. However, I am not convinced this is correct. If it isn't we may want to fix that in a future PR.

guozhangwang replied:
The `stateLock` is protecting the transactionMetadataCache, for put / get access; within the cache any in-place updates of the metadata object need to be protected by a the metadata object lock itself. The introduction of the `TransitionMetadata` is to avoid in-place updates during the prepare update period.

Let me know if you find any corner cases that will introduce a race condition.

apurvam replied:
Where is the `stateLock` acquired? Between updating the `txnMetadata` in place and actually writing the update to the log, can another request modify the metadata? If not, what locking prevents it? Or is it impossible for another request to come in at this time? 

I don't think we need to block this PR on these questions, since the pattern is everywhere. I am probably missing something in the broader locking context.

guozhangwang commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -459,8 +487,8 @@ class TransactionCoordinator(brokerId: Int,
       abortTimedOutTransactions,
-      TransactionStateManager.DefaultAbortTimedOutTransactionsIntervalMs,
-      TransactionStateManager.DefaultAbortTimedOutTransactionsIntervalMs
+      txnConfig.abortTimedOutTransactionsIntervalMs,
```
Nice catch.

guozhangwang commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -434,21 +435,48 @@ class TransactionCoordinator(brokerId: Int,
-    }
+      val coordinatorEpochAndMetadata = txnManager.getTransactionState(txnIdAndPidEpoch.transactionalId).right.flatMap {
+        case None =>
+          fatal(s"Could not find transaction metadata when trying to timeout transaction with transactionalId " +
```
Should this really be logged as a fatal error? I think it may happen that a leader migration happens between the `timedOutTransactions` call and the `foreach` loop here.

In the old code we treat such errors as a debug level logging only:

```
(error: Errors) => error match {
          case Errors.NONE =>
            info(s"Completed rollback ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} due to timeout")
          case Errors.INVALID_PRODUCER_ID_MAPPING |
               Errors.INVALID_PRODUCER_EPOCH |
               Errors.CONCURRENT_TRANSACTIONS =>
            debug(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} has aborted due to ${error.exceptionName()}")
          case e =>
            warn(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} failed due to ${error.exceptionName()}")
        })
```

apurvam replied:
The problem here is that this is happening in a background thread. If we are trying to abort a transaction but fail because the internal metadata is inconsistent, that is a severe error. And there is no way to report this back to the user apart from the logs. 

What would happen if we hit this case (which should never happen because it means internal state is inconsistent) is that transactions would never be expired, which means that read committed consumers would never make progress. If that happens, we need clues in the logs to inform us of what happens.

I was debating whether this should be at an error level, and I think that is fine as well. But it should certainly be no lower than error IMO.

apurvam replied:
I changed this to an error log level in the latest update.

guozhangwang replied:
> If we are trying to abort a transaction but fail because the internal metadata is inconsistent, that is a severe error.

`timedOutTransactions` only made a pass over the metadata and generate a candidate list of transactions that can be expired, and then in `foreach` we iterate over the list and tries to end the transaction as aborted. So if it fails to end it that is fine since the thread will add it to the candidate list again in the next call as long as the metadata is still in the cache, right?

guozhangwang commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -434,21 +435,48 @@ class TransactionCoordinator(brokerId: Int,
+
+        case Some(epochAndTxnMetadata) =>
+          if (epochAndTxnMetadata.transactionMetadata.producerId != txnIdAndPidEpoch.producerId) {
+            fatal(s"Found incorrect producerId when expiring transactionalId: ${txnIdAndPidEpoch.transactionalId}. " +
```
Ditto here.

apurvam replied:
Ditto for my response :)

guozhangwang commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -434,21 +435,48 @@ class TransactionCoordinator(brokerId: Int,
+
+      coordinatorEpochAndMetadata match {
+        case Left(error) =>
+          logger.error(s"Could not retrieve transaction metadata for transactionalId ${txnIdAndPidEpoch.transactionalId} due " +
```
Correlated to the above comment: here we are logging a second time with the error, so for each event we are logging a `fatal` then followed by an `error`.

apurvam replied:
Sure.

guozhangwang commented on core/src/test/scala/integration/kafka/api/TransactionsTest.scala:
```diff
@@ -456,6 +456,31 @@ class TransactionsTest extends KafkaServerTestHarness {
+    assertTrue(firstMessageResult.hasOffset)
+
+    // Wait for expiration cycle to kick in.
+    Thread.sleep(600)
```
Could we use a mock time for unit test? Having thread.sleep would introduce another flakiness.

apurvam replied:
This is an integration test, not a unit test. So we can't use mock time, as there are no mocks. We are actually verifying that the first record was eventually aborted and the second one was never allowed to be written from end to end (ie. producer to consumer).

Commit:
Update integration test to ensure that only one record is written and that it is eventually aborted

Commit:
Address PR comments

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -434,21 +435,49 @@ class TransactionCoordinator(brokerId: Int,
+          Left(Errors.INVALID_TXN_STATE)
+
+        case Some(epochAndTxnMetadata) =>
+          if (epochAndTxnMetadata.transactionMetadata.producerId != txnIdAndPidEpoch.producerId) {
```
Don't we need the lock to access `transactionMetadata.producerId`?

guozhangwang replied:
+1

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -434,21 +435,49 @@ class TransactionCoordinator(brokerId: Int,
-            warn(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} failed due to ${error.exceptionName()}")
-        })
-    }
+      val coordinatorEpochAndMetadata = txnManager.getTransactionState(txnIdAndPidEpoch.transactionalId).right.flatMap {
```
Why ignore the possible error in the left side?

apurvam replied:
Not sure, this is the pattern used everywhere else in this class for getting the transactionState. Perhaps @guozhangwang can comment?

apurvam replied:
To close the loop here, we synced offline and the confusion is because of the semantics of the `right.flatMap` operation. the `getTransactionState` method returns `Either[Errors, Option[CoordinatorEpochoAndMetadata]]`. If it returns the `Left` the `right.flatMap` will still return `Left[Errors]`, which is what will be stored in `coordinatorEpochAndMetadata`. So we don't need to check it explicitly since that error will be returned directly in the subsequent code.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -434,21 +435,49 @@ class TransactionCoordinator(brokerId: Int,
+              s"${epochAndTxnMetadata.transactionMetadata.producerId}")
+            Left(Errors.INVALID_PRODUCER_ID_MAPPING)
+          } else {
+            Right(epochAndTxnMetadata)
```
Seems we could take the block starting at line 461 and move it here. Not sure why we need to convert to `Either` first since we just ignore the error case anyway.

guozhangwang replied:
+1

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -434,21 +435,49 @@ class TransactionCoordinator(brokerId: Int,
+
+        case Right(epochAndMetadata) =>
+          val txnMetadata = epochAndMetadata.transactionMetadata
+          txnMetadata.inLock(txnMetadata.fenceProducerEpoch())
```
I am wondering if we need to make this a CAS to protect from an unexpected bump. Do we have any protection from this currently? I'm basically wondering what happens with a concurrent InitProducerIdRequest.

guozhangwang replied:
Hmm good point. Though even CAS would not help here, as we will check the producer epoch again in `handleEndTransaction` and `sendTxnMarkersCallback` later. So if it gets bumped again these two calls will fail.

So I think maybe we sill need to set the `pendingState` earlier than waiting for the follow-up call to abort the transaction will set its pending state. WDYT @hachikuji ?

hachikuji replied:
What I had in mind was to pass the instance of `TransactionalIdAndProducerIdEpoch` into `fenceProducerEpoch`. If either the producerId or epoch were different, then we would skip bumping the epoch. But it's a good point that we have the additional checks later which are not protected by the lock.

tedyu replied:
I noticed that producerEpoch is Short but coordinatorEpoch is Int.
Should producerEpoch be an Int as well ?

hachikuji replied:
The producer epoch is intended to be a short. This is consistent in all the request apis and the message format.

apurvam replied:
As I mentioned in my orginal comment, https://github.com/apache/kafka/pull/4137#discussion_r147026039, the locking seems suspect. There is a gap in lock acquistions which could allow race conditions. I am not sure I understood @guozhangwang's response on the other thread though, and whether this is a safe pattern. The same locking is also used in the initProducerId handling when we have to fence the producerId.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -434,21 +435,49 @@ class TransactionCoordinator(brokerId: Int,
+            txnMetadata.producerId,
+            txnMetadata.producerEpoch,
+            TransactionResult.ABORT,
+            (error: Errors) => error match {
```
You can use a pattern matching anonymous function. Just drop `(error: Errors) => error match`.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -434,21 +435,49 @@ class TransactionCoordinator(brokerId: Int,
+              case Errors.INVALID_PRODUCER_ID_MAPPING |
+                   Errors.INVALID_PRODUCER_EPOCH |
+                   Errors.CONCURRENT_TRANSACTIONS =>
+                debug(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} has aborted due to ${error.exceptionName()}")
```
nit: drop parenthesis from `exceptionName`. One below too.

hachikuji reviewed (commented):
Thanks for the investigation and the patch. Left a few comments.

Commit:
Address PR comments

tedyu commented:
@apurvam 
Do you think your PR would also fix this failing test ?

https://builds.apache.org/job/kafka-pr-jdk9-scala2.12/2293/testReport/junit/kafka.api/TransactionsTest/testFencingOnCommit/

apurvam commented:
no that test doesn't look related at all, @tedyu

Commit:
Fix the race condition in handleInitProducerId and
abortTimedoutTransaction by introducing a new PrepareEpochFence state
which will lock out other requests for the same transactionalId until
finally transitioning to the PrepareAbort state.

Commit:
Lock txnMetadata when checking if the producerId has changed

Commit:
Minor cleanups

guozhangwang commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -308,16 +309,24 @@ class TransactionCoordinator(brokerId: Int,
+            else if (producerEpoch < txnMetadata.producerEpoch)
               Left(Errors.INVALID_PRODUCER_EPOCH)
-            else if (txnMetadata.pendingTransitionInProgress)
+            else if (txnMetadata.pendingTransitionInProgress && txnMetadata.pendingState.isDefined && txnMetadata.pendingState.get != PrepareEpochFence)
```
The first two conditions are the same?

apurvam commented:
@hachikuji @guozhangwang all the tests passed. I also took care of all the comments, and fixed the existing race condition in `handleInitProducerId` as well. Essentially I introduced a new pending state `PrepareEpochFence` which would lock out any other operation on the transactional id until the fencing operation completes. This makes the logic much tighter IMO.

guozhangwang commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -450,23 +459,27 @@ class TransactionCoordinator(brokerId: Int,
-              })
-            Right(epochAndTxnMetadata)
-          }
+            if (txnMetadata.pendingTransitionInProgress) {
```
I think we should have the inLock covering the first condition as well since it is reading the pending state right?

guozhangwang commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -434,20 +445,46 @@ class TransactionCoordinator(brokerId: Int,
+              s"${epochAndTxnMetadata.transactionMetadata.producerId}")
+            Left(Errors.INVALID_PRODUCER_ID_MAPPING)
+          } else {
+            if (txnMetadata.pendingTransitionInProgress) {
```
This check is not covered by the object lock, is that intentional?

apurvam replied:
fixed.

Commit:
Check if there is a pending state change inside the lock

Commit:
Tighten up locking

Commit:
Minor cleanup of return value

guozhangwang commented:
Made several passes and synced offline with @apurvam on a few comments. It LGTM now, merging to trunk and 1.0.

cc @hachikuji .

ijuma commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -307,16 +309,25 @@ class TransactionCoordinator(brokerId: Int,
                   PrepareAbort
+
+                if (nextState == PrepareAbort && txnMetadata.pendingState.isDefined
+                  && txnMetadata.pendingState.get == PrepareEpochFence) {
```
FYI, you could do `txnMetadata.pendingState == Some(PreparedEpochFence)` to combine two checks into one (and avoid `get`, which is generally a code smell).

ijuma commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -1382,12 +1382,24 @@ object TestUtils extends Logging {
   }
 
-  def createTransactionalProducer(transactionalId: String, servers: Seq[KafkaServer], batchSize: Int = 16384) = {
+  def consumeRemainingRecords[K, V](consumer: KafkaConsumer[K, V], timeout: Long): Seq[ConsumerRecord[K, V]] = {
```
Nit: I think this should be `consumeRecordsFor` and `timeout` should be `duration`. Assuming I understood the intent (consume records until `timeout/duration` expires). And we should probably have scaladoc explaining this.

apurvam replied:
fixed in #4146

ijuma reviewed (commented):
I was trying to understand this change and left a couple of minor comments. I think the test utility name is a bit confusing and is worth addressing (if I understood it correctly).

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -212,7 +213,8 @@ class TransactionCoordinator(brokerId: Int,
           Right(coordinatorEpoch, txnMetadata.prepareFenceProducerEpoch())
-        case Dead =>
+
+        case Dead | PrepareEpochFence =>
```
I'm not sure I understand why this is an illegal state. Isn't it possible for a transaction to be aborted at about the same time that a new producer sends InitProducerId?

apurvam replied:
When we fence the epoch, we immediately set the `txnMetadata.pendingState` to `PrepareEpochFence`. This locks all other requests, since any new request which comes in when the pending state is defined will be rejected with the `CONCURRENT_TRANSACTIONS` error code.

Once we are in this pending state, we immediately begin ending the transaction, in which case the pending state becomes `PrepareAbort`. We then follow the normal state transitions for the end transaction process, which ultimately ends in `CompleteAbort`. 

So at no time do we ever complete the transition to `PrepareEpochFence`. So if the transaciton metadata is in this state, it is always illegal.

hachikuji replied:
@guozhangwang explained offline. The `PrepareEpochFence` is special in the sense that you can't actually transition to it. This is quite confusing, but I guess the check makes sense.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -434,20 +445,52 @@ class TransactionCoordinator(brokerId: Int,
+          if (producerIdHasChanged) {
+            error(s"Found incorrect producerId when expiring transactionalId: ${txnIdAndPidEpoch.transactionalId}. " +
+              s"Expected producerId: ${txnIdAndPidEpoch.producerId}. Found producerId: " +
+              s"${epochAndTxnMetadata.transactionMetadata.producerId}")
```
We're still accessing this outside the lock. Also, we could use `txnMetadata`.

apurvam replied:
Yes. I realized this last night, will submit a follow up to fix this particular bug.

apurvam replied:
see #4146

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -434,20 +445,52 @@ class TransactionCoordinator(brokerId: Int,
-          case e =>
-            warn(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} failed due to ${error.exceptionName()}")
-        })
+      txnManager.getTransactionState(txnIdAndPidEpoch.transactionalId).right.flatMap {
```
Instead of `flatMap`, you can use `foreach`. Then we don't need the pointless `Either` results.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -434,20 +445,52 @@ class TransactionCoordinator(brokerId: Int,
 
   private def abortTimedOutTransactions(): Unit = {
     txnManager.timedOutTransactions().foreach { txnIdAndPidEpoch =>
```
Looking over this code, `timedOutTransactions` seems to access internal `TransactionMetadata` state without grabbing the lock.

hachikuji reviewed (commented):
Left a few additional comments. I started a follow-up patch which I will submit shortly.
